pmd-core/src/main/java/net/sourceforge/pmd/RuleSet.java

```diff
@@ -493,18 +493,19 @@ public void start(RuleContext ctx) {
      *            the current context
      */
     public void apply(List<? extends Node> acuList, RuleContext ctx) {
-        long start = System.nanoTime();
+        TimeTracker.startOperation(TimedOperationCategory.RULE);
         for (Rule rule : rules) {
             try {
                 if (!rule.isRuleChain() && applies(rule, ctx.getLanguageVersion())) {
+                    TimeTracker.startOperation(TimedOperationCategory.RULE, rule.getName());
                     rule.apply(acuList, ctx);
-                    long end = System.nanoTime();
-                    Benchmarker.mark(Benchmark.Rule, rule.getName(), end - start, 1);
-                    start = end;
+                    TimeTracker.finishOperation();
                 }
             } catch (RuntimeException e) {
                 if (ctx.isIgnoreExceptions()) {
+                    TimeTracker.finishOperation();
```

Comment:
This won't work right when `ctx.isIgnoreException` is false.  It's _also_ not guaranteed to be paired with the appropriate `startOperation` if the exception is from the `if` conditional on line 499 immediately above.

Nice attempt, at least on the right track!  Bullet proof exception handling isn't always easy.  It's for this reason I find the try-with-resouces style to be so appealing, such flow-logic bugs aren't possible.

This is probably where the original code was loosing time for me in regression testing!